src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReadBufferState.cs

```diff
@@ -2,32 +2,168 @@
+            while (_count < _buffer.Length);
+
+            ProcessReadBytes();
+        }
+
+        /// <summary>
+        /// Advances the buffer in anticipation of a subsequent read operation.
+        /// </summary>
+        public void AdvanceBuffer(int bytesConsumed)
+        {
+            Debug.Assert(bytesConsumed <= _count);
+            Debug.Assert(!_isFinalBlock || _count == bytesConsumed, "The reader should have thrown if we have remaining bytes.");
+
+            _count -= bytesConsumed;
+
+            if (!_isFinalBlock)
+            {
+                // Check if we need to shift or expand the buffer because there wasn't enough data to complete deserialization.
+                if ((uint)_count > ((uint)_buffer.Length / 2))
+                {
+                    // We have less than half the buffer available, double the buffer size.
+                    byte[] oldBuffer = _buffer;
+                    int oldMaxCount = _maxCount;
+                    byte[] newBuffer = ArrayPool<byte>.Shared.Rent((_buffer.Length < (int.MaxValue / 2)) ? _buffer.Length * 2 : int.MaxValue);
+
+                    // Copy the unprocessed data to the new buffer while shifting the processed bytes.
+                    Buffer.BlockCopy(oldBuffer, _offset + bytesConsumed, newBuffer, 0, _count);
+                    _buffer = newBuffer;
+                    _offset = 0;
+                    _maxCount = _count;
+
+                    // Clear and return the old buffer
+                    new Span<byte>(oldBuffer, 0, oldMaxCount).Clear();
+                    ArrayPool<byte>.Shared.Return(oldBuffer);
+                }
+                else if (_count != 0)
+                {
+                    // Shift the processed bytes to the beginning of buffer to make more room.
+                    Buffer.BlockCopy(_buffer, _offset + bytesConsumed, _buffer, 0, _count);
+                    _offset = 0;
+                }
+            }
+        }
+
+        private void ProcessReadBytes()
+        {
+            if (_count > _maxCount)
+            {
+                _maxCount = _count;
+            }
+
+            if (_isFirstBlock)
+            {
+                _isFirstBlock = false;
+
+                // Handle the UTF-8 BOM if present
+                Debug.Assert(_buffer.Length >= JsonConstants.Utf8Bom.Length);
+                if (_buffer.AsSpan(0, _count).StartsWith(JsonConstants.Utf8Bom))
+                {
+                    _offset = (byte)JsonConstants.Utf8Bom.Length;
+                    _count -= JsonConstants.Utf8Bom.Length;
+                }
+            }
         }
 
         public void Dispose()
         {
             // Clear only what we used and return the buffer to the pool
-            new Span<byte>(Buffer, 0, ClearMax).Clear();
+            new Span<byte>(_buffer, 0, _maxCount).Clear();
 
-            byte[] toReturn = Buffer;
-            Buffer = null!;
+            byte[] toReturn = _buffer;
+            _buffer = null!;
```

Comment:
considering you return the copy of this buffer state in some of the methods is it safe to do it like this? Should you add some guarding against double free?

Reply:
> considering you return the copy of this buffer state in some of the methods is it safe to do it like this? Should you add some guarding against double free?

Yes, assuming you somehow end up disposing two copies of the same struct then that is a possibility, but we're not doing this in the handful of methods that creates the struct. 

> perhaps some #if DEBUG extra code protecting us against this pattern (i.e. wrap buffer in ref type, set it to null on Dispose and check with Debug.Assert)

Can you give an example? Not sure I understand. FWIW this is not new code, I'm just moving existing logic to the struct.

Reply:
```csharp
ReadBufferState a = ...;
ReadBufferState b = a;
a.Dispose();
b.Dispose(); // double free
```

Reply:
I understand the condition, my question is what assert could guard against such a potential scenario.